### PR TITLE
CLOUDP-80517: error and update int test for AtlasProject

### PR DESCRIPTION
**Pull request by antonlisovenko**
Extended the integration test for update and error handling.
Update part is with TODO as we need a more deterministic approach of knowing when the Operator has started working on the changes (https://jira.mongodb.org/browse/CLOUDP-80550)

Output:
```
=== RUN   TestAPIs
Running Suite: Project Controller Suite
=======================================
Random Seed: 1610734310
Will run 3 of 3 specs

STEP: bootstrapping test environment
2021-01-15T18:11:50.426Z	DEBUG	controller-runtime.test-env	starting control plane	{"api server flags": []}
2021-01-15T18:11:54.607Z	DEBUG	controller-runtime.test-env	installing CRDs
2021-01-15T18:11:54.607Z	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "../../../config/crd/bases"}
2021-01-15T18:11:54.610Z	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "atlas.mongodb.com_atlasclusters.yaml"}
2021-01-15T18:11:54.634Z	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "atlas.mongodb.com_atlasprojects.yaml"}
2021-01-15T18:11:54.651Z	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "atlasclusters.atlas.mongodb.com"}
2021-01-15T18:11:54.684Z	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "atlasprojects.atlas.mongodb.com"}
2021-01-15T18:11:54.774Z	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "atlas.mongodb.com/v1"}
2021-01-15T18:11:54.774Z	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "atlas.mongodb.com/v1"}
2021-01-15T18:11:54.880Z	DEBUG	controller-runtime.test-env	installing webhooks
2021-01-15T18:11:55.312Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
AtlasProject Creating the project 
  Should Succeed
  /Users/alisovenko/workspace/mongodb-atlas-kubernetes/test/int/project/project_test.go:68
STEP: Creating the namespace 
2021-01-15T18:11:55.312Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2021-01-15T18:11:55.312Z	INFO	controller	Starting EventSource	{"reconcilerGroup": "atlas.mongodb.com", "reconcilerKind": "AtlasProject", "controller": "atlasproject", "source": "kind source: /, Kind="}
STEP: Creating the Secret testct96n/my-atlas-key
2021-01-15T18:11:55.413Z	INFO	controller	Starting Controller	{"reconcilerGroup": "atlas.mongodb.com", "reconcilerKind": "AtlasProject", "controller": "atlasproject"}
2021-01-15T18:11:55.413Z	INFO	controller	Starting workers	{"reconcilerGroup": "atlas.mongodb.com", "reconcilerKind": "AtlasProject", "controller": "atlasproject", "worker count": 1}
2021-01-15T18:11:55.413Z	INFO	controllers.AtlasProject	-> Starting AtlasProject reconciliation	{"atlasproject": "testct96n/test-project", "spec": {"name":"Test Project","connectionSecretRef":{"name":"my-atlas-key"}}}
2021-01-15T18:11:55.413Z	INFO	controllers.AtlasProject	Reading Atlas API credentials from the AtlasProject Secret testct96n/my-atlas-key	{"atlasproject": "testct96n/test-project"}
2021-01-15T18:11:56.136Z	DEBUG	controllers.AtlasProject	HTTP Request (GET) https://cloud-qa.mongodb.com/api/atlas/v1.0/groups/byName/Test%20Project [time (ms): 618, status: 401]	{"atlasproject": "testct96n/test-project"}
2021-01-15T18:11:58.341Z	DEBUG	controllers.AtlasProject	HTTP Request (POST) https://cloud-qa.mongodb.com/api/atlas/v1.0/groups [time (ms): 2203, status: 201]	{"atlasproject": "testct96n/test-project"}
2021-01-15T18:11:58.341Z	INFO	controllers.AtlasProject	Created Atlas Project	{"atlasproject": "testct96n/test-project", "name": "Test Project", "id": "6001daec14ba3b0d87e7920f"}
2021-01-15T18:11:58.341Z	INFO	controllers.AtlasProject	Status update	{"atlasproject": "testct96n/test-project", "lastCondition": {"type":"IPAccessListReady","status":"False","lastTransitionTime":null}}
2021-01-15T18:11:58.345Z	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "atlas.mongodb.com", "reconcilerKind": "AtlasProject", "controller": "atlasproject", "name": "test-project", "namespace": "testct96n"}
STEP: Removing Atlas Project 6001daec14ba3b0d87e7920f
STEP: Removing the namespace testct96n
•
------------------------------
AtlasProject Creating the project 
  Should fail if Secret is wrong
  /Users/alisovenko/workspace/mongodb-atlas-kubernetes/test/int/project/project_test.go:82
STEP: Creating the namespace 
STEP: Creating the Secret testx2fcp/my-atlas-key
2021-01-15T18:11:59.360Z	INFO	controllers.AtlasProject	-> Starting AtlasProject reconciliation	{"atlasproject": "testx2fcp/test-project", "spec": {"name":"Test Project","connectionSecretRef":{"name":"non-existent-secret"}}}
2021-01-15T18:11:59.360Z	INFO	controllers.AtlasProject	Reading Atlas API credentials from the AtlasProject Secret testx2fcp/non-existent-secret	{"atlasproject": "testx2fcp/test-project"}
2021-01-15T18:11:59.360Z	INFO	controllers.AtlasProject	Status update	{"atlasproject": "testx2fcp/test-project", "lastCondition": {"type":"ProjectReady","status":"False","lastTransitionTime":null,"reason":"AtlasCredentialsNotProvided","message":"Secret \"non-existent-secret\" not found"}}
STEP: Removing the namespace testx2fcp
•
------------------------------
AtlasProject Updating the project 
  Should Succeed
  /Users/alisovenko/workspace/mongodb-atlas-kubernetes/test/int/project/project_test.go:100
STEP: Creating the namespace 
STEP: Creating the Secret testqtqzp/my-atlas-key
STEP: Creating the project first
2021-01-15T18:12:00.655Z	INFO	controllers.AtlasProject	-> Starting AtlasProject reconciliation	{"atlasproject": "testqtqzp/test-project", "spec": {"name":"Test Project","connectionSecretRef":{"name":"my-atlas-key"}}}
2021-01-15T18:12:00.655Z	INFO	controllers.AtlasProject	Reading Atlas API credentials from the AtlasProject Secret testqtqzp/my-atlas-key	{"atlasproject": "testqtqzp/test-project"}
2021-01-15T18:12:00.877Z	DEBUG	controllers.AtlasProject	HTTP Request (GET) https://cloud-qa.mongodb.com/api/atlas/v1.0/groups/byName/Test%20Project [time (ms): 221, status: 401]	{"atlasproject": "testqtqzp/test-project"}
2021-01-15T18:12:03.880Z	DEBUG	controllers.AtlasProject	HTTP Request (POST) https://cloud-qa.mongodb.com/api/atlas/v1.0/groups [time (ms): 3003, status: 201]	{"atlasproject": "testqtqzp/test-project"}
2021-01-15T18:12:03.880Z	INFO	controllers.AtlasProject	Created Atlas Project	{"atlasproject": "testqtqzp/test-project", "name": "Test Project", "id": "6001daf114ba3b0d87e7923a"}
2021-01-15T18:12:03.880Z	INFO	controllers.AtlasProject	Status update	{"atlasproject": "testqtqzp/test-project", "lastCondition": {"type":"IPAccessListReady","status":"False","lastTransitionTime":null}}
2021-01-15T18:12:03.882Z	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "atlas.mongodb.com", "reconcilerKind": "AtlasProject", "controller": "atlasproject", "name": "test-project", "namespace": "testqtqzp"}
STEP: Updating the project (the existing project is expected to be read)
2021-01-15T18:12:04.675Z	INFO	controllers.AtlasProject	-> Starting AtlasProject reconciliation	{"atlasproject": "testqtqzp/test-project", "spec": {"name":"Test Project","connectionSecretRef":{"name":"my-atlas-key"},"projectIpAccessList":[{"cidrBlock":"0.0.0.0/0"}]}}
2021-01-15T18:12:04.675Z	INFO	controllers.AtlasProject	Reading Atlas API credentials from the AtlasProject Secret testqtqzp/my-atlas-key	{"atlasproject": "testqtqzp/test-project"}
2021-01-15T18:12:04.905Z	DEBUG	controllers.AtlasProject	HTTP Request (GET) https://cloud-qa.mongodb.com/api/atlas/v1.0/groups/byName/Test%20Project [time (ms): 230, status: 200]	{"atlasproject": "testqtqzp/test-project"}
2021-01-15T18:12:04.905Z	INFO	controllers.AtlasProject	Status update	{"atlasproject": "testqtqzp/test-project", "lastCondition": {"type":"IPAccessListReady","status":"False","lastTransitionTime":null}}
2021-01-15T18:12:04.908Z	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "atlas.mongodb.com", "reconcilerKind": "AtlasProject", "controller": "atlasproject", "name": "test-project", "namespace": "testqtqzp"}
2021-01-15T18:12:09.386Z	INFO	controllers.AtlasProject	-> Starting AtlasProject reconciliation	{"atlasproject": "testx2fcp/test-project", "spec": {"name":"Test Project","connectionSecretRef":{"name":"non-existent-secret"}}}
2021-01-15T18:12:09.386Z	INFO	controllers.AtlasProject	Reading Atlas API credentials from the AtlasProject Secret testx2fcp/non-existent-secret	{"atlasproject": "testx2fcp/test-project"}
2021-01-15T18:12:09.386Z	INFO	controllers.AtlasProject	Status update	{"atlasproject": "testx2fcp/test-project", "lastCondition": {"type":"ProjectReady","status":"False","lastTransitionTime":null,"reason":"AtlasCredentialsNotProvided","message":"Secret \"non-existent-secret\" not found"}}
STEP: Removing Atlas Project 6001daf114ba3b0d87e7923a
STEP: Removing the namespace testqtqzp

• [SLOW TEST:14.953 seconds]
AtlasProject
/Users/alisovenko/workspace/mongodb-atlas-kubernetes/test/int/project/project_test.go:23
  Updating the project
  /Users/alisovenko/workspace/mongodb-atlas-kubernetes/test/int/project/project_test.go:99
    Should Succeed
    /Users/alisovenko/workspace/mongodb-atlas-kubernetes/test/int/project/project_test.go:100
------------------------------
STEP: tearing down the test environment
E0115 18:12:15.585707   61579 reflector.go:383] pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125: Failed to watch *v1.Secret: Get "http://127.0.0.1:64196/api/v1/secrets?allowWatchBookmarks=true&resourceVersion=62&timeoutSeconds=506&watch=true": dial tcp 127.0.0.1:64196: connect: connection refused
E0115 18:12:15.585707   61579 reflector.go:383] pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125: Failed to watch *v1.AtlasProject: Get "http://127.0.0.1:64196/apis/atlas.mongodb.com/v1/atlasprojects?allowWatchBookmarks=true&resourceVersion=68&timeoutSeconds=427&watch=true": dial tcp 127.0.0.1:64196: connect: connection refused


Ran 3 of 3 Specs in 25.204 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestAPIs (25.21s)
PASS
```

### CODE REVIEW

File: pkg/api/v1/status/condition.go [hunk 1]
```diff
@@ -66,7 +66,22 @@ func TrueCondition(conditionType ConditionType) Condition {
...
 		Status:             corev1.ConditionTrue,
 		LastTransitionTime: metav1.Now(),
-		Reason:             "",
-		Message:            "",
 	}
 }
+
+// FalseCondition returns the COndition that has the 'Status' set to 'false' and 'Type' to 'conditionType'.
+// The reason and message can be provided optionally
+func FalseCondition(conditionType ConditionType, reasonAndMessage ...string) Condition {
```
Comment (vasilevp) [inline pkg/api/v1/status/condition.go:74]: [nit]
That looks OK for tests, but in actual code we may want to preserve the `workflow.Reason` type (i.e. by using `conditions.New(type, false).WithReason(reason)` or something of that sort)

File: pkg/api/v1/status/condition.go [hunk 2]
```diff
@@ -66,7 +66,22 @@ func TrueCondition(conditionType ConditionType) Condition {
 		Type:               conditionType,
 		Status:             corev1.ConditionTrue,
 		LastTransitionTime: metav1.Now(),
-		Reason:             "",
-		Message:            "",
 	}
 }
+
+// FalseCondition returns the COndition that has the 'Status' set to 'false' and 'Type' to 'conditionType'.
```
Comment (vasilevp) [inline pkg/api/v1/status/condition.go:72]: [nit] typo in **CO**ndition

**Review by vasilevp — APPROVED** [review]
No blockers from me

File: pkg/api/v1/status/condition.go (hunk 1 above)
Comment (antonlisovenko) [reply pkg/api/v1/status/condition.go:74]: I think you are right, I was too lazy and decided to postpone `with` solution for later.
I'll address this in the next PR which is dependent on this one and is on-going.
thanks 👍

File: pkg/api/v1/status/condition.go (hunk 2 above)
Comment (antonlisovenko) [reply pkg/api/v1/status/condition.go:72]: will fix in the current PR that is based on the current one

**Pull request merged**